pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/AutoRebalancePartitionAssignmentGenerator.java

```diff
@@ -0,0 +1,113 @@
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.controller.helix.core.realtime;
+
+import com.linkedin.pinot.common.config.RoutingConfig;
+import com.linkedin.pinot.common.config.TableConfig;
+import com.linkedin.pinot.common.utils.CommonConstants.Helix.DataSource.RoutingTableBuilderName;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.controller.rebalancer.strategy.AutoRebalanceStrategy;
+
+
+/**
+ * Generates partition assignment for all tables provided, using AutoRebalanceStrategy
+ */
+public class AutoRebalancePartitionAssignmentGenerator extends PartitionAssignmentGenerator {
+
+  public AutoRebalancePartitionAssignmentGenerator(TableConfig tableConfig, int nKafkaPartitions,
+      List<String> instanceNames, Map<String, TableConfig> allTableConfigsInTenant,
+      Map<String, ZNRecord> currentPartitionAssignment) {
+
+    super(tableConfig, nKafkaPartitions, instanceNames, currentPartitionAssignment);
+
+    _tablesForPartitionAssignment = new ArrayList<>();
+    for (TableConfig config : allTableConfigsInTenant.values()) {
+      RoutingConfig routingConfig = config.getRoutingConfig();
+      if (routingConfig == null || routingConfig.getRoutingTableBuilderName() == null ||
+          !RoutingTableBuilderName.PartitionAwareRealtime.toString().equals(routingConfig.getRoutingTableBuilderName())) {
+        _tablesForPartitionAssignment.add(config.getTableName());
+      }
+    }
+  }
+
+  @Override
+  public Map<String, ZNRecord> generatePartitionAssignment() {
+    Map<String, ZNRecord> tableNameToPartitionAssignment = new HashMap<>(_tablesForPartitionAssignment.size());
+    final String partitionJoiner = "_";
+
+    Map<String, Map<String, String>> currentPartitions = new HashMap<>();
+    List<String> newPartitions = new ArrayList<>();
+    for (String realtimeTableName : _tablesForPartitionAssignment) {
+
+      // construct current partitions map
+      ZNRecord partitionAssignmentZNode = _currentPartitionAssignment.get(realtimeTableName);
+      if (partitionAssignmentZNode != null) {
+
+        Map<String, List<String>> partitionToInstances = partitionAssignmentZNode.getListFields();
+        for (Map.Entry<String, List<String>> partitionEntry : partitionToInstances.entrySet()) {
+          String partitionNumber = partitionEntry.getKey();
+          List<String> partitionInstances = partitionEntry.getValue();
+          String key = realtimeTableName + partitionJoiner + partitionNumber;
```

Comment:
create a class to represent RealtimePartition which will encapsulate the name generation logic. Current Assignment could be Map<RealtimePartition, List<RealtimeInstance>>

Reply:
I can create a RealtimePartition class, but would prefer to keep it private to just the AutoRebalancePartitionAssignment, as these partition names make sense only for this strategy.